src/transformStateWithClones.js

```diff
@@ -7,7 +7,26 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    let newState = { ...currentState };
+
+    if (action.type === 'addProperties') {
+      Object.assign(newState, action.extraData);
+    } else if (action.type === 'removeProperties') {
+      action.keysToRemove.forEach((key) => {
+        delete newState[key];
+      });
+    } else if (action.type === 'clear') {
+      newState = {};
+    }
+    history.push({ ...newState });
+    currentState = { ...newState };
```

Comment:
Consider using a `switch` statement instead of multiple `if...else if` statements for better readability and to adhere to the checklist item regarding the use of `switch` when you have a limited number of conditions.